File: app/src/main/java/com/aas/moviecatalog/adapter/MovieAdapter.java
package com.aas.moviecatalog.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.aas.moviecatalog.R;
import com.aas.moviecatalog.api.TheMovieDBAPI;
import com.aas.moviecatalog.model.Movie;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieViewHolder> {
    private Context context;
    private ArrayList<Movie> listMovie;
    private MovieListener movieListener;

    public void setListMovie(ArrayList<Movie> listMovie) {
        this.listMovie = listMovie;
        notifyDataSetChanged();
    }

    public MovieAdapter(Context context, MovieListener movieListener) {
        this.context = context;
        this.movieListener = movieListener;
        listMovie = new ArrayList<>();
    }

    @NonNull
    @Override
    public MovieViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.list_item, parent, false);

        return new MovieViewHolder(view, movieListener);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieViewHolder holder, int position) {
        holder.bind(listMovie.get(position));

    }

    @Override
    public int getItemCount() {
        return listMovie.size();
    }

    public class MovieViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView tvJudul;
        ImageView ivGambar;
        MovieListener movieListener;

        MovieViewHolder(@NonNull View itemView, MovieListener movieListener) {
            super(itemView);

            tvJudul = itemView.findViewById(R.id.tvJudul);
            ivGambar = itemView.findViewById(R.id.ivGambar);
            this.movieListener = movieListener;

            itemView.setOnClickListener(this);
        }

        void bind(Movie movie){
            tvJudul.setText(movie.getTitle());
            Picasso.get().load(TheMovieDBAPI.getPoster(movie.getPosterPath())).into(ivGambar);

        }

        @Override
        public void onClick(View view) {
            movieListener.onMovieItemClick(getAdapterPosition());
        }
    }
    public interface MovieListener{
        void onMovieItemClick(int position);
    }
}

File: app/src/main/java/com/aas/moviecatalog/movie/MovieAdapter.kt
package com.aas.moviecatalog.movie

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.aas.moviecatalog.R
import com.aas.moviecatalog.api.ApiRepository
import com.aas.moviecatalog.model.Movie
import com.squareup.picasso.Picasso
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.list_item.view.*

class MovieAdapter(
    private val context: Context,
    private val movies: MutableList<Movie>,
    private val onClickListener: (Movie) -> Unit
) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {

    fun setMovie(movie: List<Movie>) {
        movies.clear()
        movies.addAll(movie)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(
        LayoutInflater.from(context).inflate(R.layout.list_item, parent, false)
    )

    override fun getItemCount(): Int = movies.size

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(context, movies[position], onClickListener)
    }

    class ViewHolder(override val containerView: View) : RecyclerView.ViewHolder(containerView),
        LayoutContainer {
        fun bind(context: Context, movie: Movie, onClickListener: (Movie) -> Unit) {
            itemView.tvJudul.text = movie.title
            itemView.tvOverview.text = movie.overview
            Picasso
                .get()
                .load(ApiRepository.BASE_IMAGE_URL + movie.poster_path)
                .into(itemView.ivGambar)
            containerView.setOnClickListener { onClickListener(movie) }
        }
    }
}